File: Python/1937_욕심쟁이 판다.py
import sys
import heapq as h

if __name__ == '__main__':
	n = int(sys.stdin.readline())
	bamboo = [list(map(int,sys.stdin.readline().split())) for _ in range(n)]
	day = [[1]*n for _ in range(n)] # 해당 좌표에서 살 수 있는 일수(현재 좌표의 대나무만 먹을 수 있는 경우인 1로 초기화)
	q = [] # priority queue를 위한 배열
	for i in range(n):
		for j in range(n):
			h.heappush(q, (-bamboo[i][j],[i,j])) # 대나무의 수가 큰 것부터 내림차순으로 좌표를 정렬하기 위해 대나무의수*-1로 정렬 값 설정(heapq를 사용하면 오름차순으로 정렬하기 때문)
	
	K = 0 # 판다가 최대한 살 수 있는 일수
	dir = [[-1,0],[1,0],[0,-1],[0,1]] # 상하좌우
	while q:
		x, y = h.heappop(q)[1] # 대나무의 수가 많은 좌표부터 pop하여 탐색
		
		for dx, dy in dir: # 상하좌우
			dx += x
			dy += y # 좌표 값 계산
			if 0 <= dx < n and 0 <= dy < n: # 범위 체크
				if bamboo[x][y] < bamboo[dx][dy]: # 이동할 좌표가 현재 좌표보다 더 많은 대나무를 가지고 있을 경우
					day[x][y] = max(day[x][y], day[dx][dy]+1) # 이동할 좌표의 수명 + 1과 현재 좌표의 수명을 비교하여 더 큰 값 저장
		K = max(K, day[x][y]) # 판다가 최대한 살 수 있는 일수 저장을 위한 비교
		
	print(K)

File: C++/1937_욕심쟁이판다.cpp
#include <iostream>
#include <queue>
#include <algorithm>
using namespace std;

struct Pos{
	int n_bamboo, x, y;
}; // 대나무의 수, 좌표 값

bool operator<(Pos a, Pos b){
	return a.n_bamboo < b.n_bamboo;
} // priority queue를 대나무 수의 내림차순을 기준으로 정렬하기 위함

int main() {
	int n;
	cin >> n;
	
	priority_queue<Pos> q; // priority queue
	int **bamboo = new int*[n]; // 대나무의 수를 저장할 배열
	int **day = new int*[n]; // 현재 위치에서 팬더가 살 수 있는 최대 일수를 저장하기 위한 배열
	for(int i = 0; i < n; i++){
		bamboo[i] = new int[n];
		day[i] = new int[n];
		for(int j = 0; j < n; j++){
			cin >> bamboo[i][j];
			q.push({bamboo[i][j],i,j});
			day[i][j] = 1; // 일수는 1로 초기화(현재 위치에서만 살 수 있을 때)
		}
	}
	
	int dir[4][2] = {{-1,0},{1,0},{0,-1},{0,1}}; // 상하좌우
	int lifespan = 0; // 판다가 최대한 살 수 있는 일수
	while(!q.empty()){ // 대나무의 수가 가장 많은 좌표부터 탐색 진행
		int x = q.top().x, y = q.top().y;
		q.pop(); // priority queue의 가장 앞쪽 좌표 pop(대나무의 수가 가장 많은 좌표)
		
		for(int d = 0; d < 4; d++){ // 상하좌우 이동
			int dx = x+dir[d][0], dy = y+dir[d][1]; // 이동 후 좌표 계산
			if(dx >= 0 && dx < n && dy >= 0 && dy < n){ // 범위 체크
				if(bamboo[x][y] < bamboo[dx][dy]){ // 현재 좌표보다 이동할 좌표가 더 큰 좌표라면
					day[x][y] = max(day[x][y],day[dx][dy]+1); // 현재 저장된 생존 일수와 이동좌표의 생존일수 +1을 비교하여 더 큰 값 저장
				}
			}
		}
		lifespan = max(lifespan,day[x][y]); // 계산 후, 판다가 최대한 살 수 있는 일수 비교 후 저장
	}
	
	cout << lifespan;
	
	return 0;
}
